### ISSUE TITLE
write extractor functions fixef, ranef, vcov, logLik

### ISSUE THREAD

**Issue by mebrooks** [glmmTMB/glmmTMB#25]
as.function () to return the deviance is helpful

**Comment by mebrooks** [comment]
a getME() function could make lme4 compatibility easy

**Comment by mmaechler** [comment]
`fixef()` has been implemented, returning an class "fixef.glmmTMB", a list
and a (too ?) smart print method :

```
> fm0 <- glmmTMB(Reaction ~ 1 + ( 1 | Subject), sleepstudy)
> fixef(fm0)
(Intercept)
   298.5079

## but then, it only *prints* as if it was a named numeric vector; it ain't at all -- hmm....

> str( fixef(fm0) )
List of 2
 $ conditional_model: Named num 299
  ..- attr(*, "names")= chr "(Intercept)"
 $ zero_inflation   : num(0)
 - attr(*, "class")= chr "fixef.glmmTMB"
```

and I wonder if this has all been discussed.

Here's a  **critical review**, i.e., a request for comments:

* It's very good to return a list with names.

* The names are bit too long IMO; users will use [[1]] or [[2]] instead which
  defeats part of the purpose of names.
  Couldn't we, e.g.,  use `CM` and `ZI` ?

* The `print()` method should be less sneeky tricky and *not* suggest that
  the fixef() object is just named numeric vector.
  It could instead print something like
```
Conditional Model:
(Intercept)
   298.5079
Zero Inflation: none
```

**Comment by bbolker** [comment]
agree with @mmaechler .  For the names, we should try to come up with consistent user-facing abbreviations for the conditional, zero-inflated, and dispersion parameters (which may or may not match with the internal abbreviations). I would prefer `cond`, `zi`, `disp` to `CM`, `ZI` (I'm a little surprised that @mmaechler is recommending uppercase here ...)

**Comment by bbolker** [comment]
note that `as.function` should return (something like) the deviance *function*, not the deviance ... it might just return `obj` (some difficulty with whether it returns `obj$fn` or `obj$gr` or a wrapper function that returns the negative log-likelihood or deviance with the gradient as an attribute ... ?)

**Comment by koenvanbenthem** [comment]
I agree on that and the future print function will be similar to what you
describe. The current fixef print method was - as far as I know, but I
didn't write it - mainly implemented to have a quick look at them. The code
I had written for them (but not uploaded to avoid conflicting behaviour),
was:
print.fixef.glmmTMB<-function(object, digits = max(3, getOption("digits") -
3),
                        correlation = NULL, symbolic.cor = FALSE,
                        signif.stars = getOption("show.signif.stars"),
                        ranef.comp = "Std.Dev.", ...){

  if(length(object$conditional_model)>0){
    cat("\nConditional model\n",fill=TRUE)
    print.default(format(object$conditional_model,digits=digits), print.gap
= 2L, quote = FALSE,...)
  }
  if(length(object$zero_inflation)>0){
    cat("\nZero inflation\n",fill=TRUE)
    print.default(format(object$zero_inflation,digits=digits), print.gap =
2L, quote = FALSE,...)
  }
}



2015-08-25 16:08 GMT+02:00 Martin Mächler <notifications@github.com>:

**Comment by bbolker** [comment]
looks good 

* names of list need to be changed consistently in `fixef.glmmTMB` and `print.fixef.glmmTMB`.
* with the possible exception of `digits`, I think the additional arguments you have here (carried over from `print.ranef.glmmTMB` or `print.summary.glmmTMB`) are useless in this context.

**Comment by mmaechler** [comment]
@bbolker -- for two letter names which **are** acronyms, I do like two uppercase letters, because then it suggests that I should de-acronymize it to "read"...  but I have no strong opinion, and we should really settle now on a common scheme... we need the same list naming -- only for ("cond", "zi") or
("CM", "ZI")  also for `VarCorr()`.

**Comment by mebrooks** [comment]
Koen and I did discuss the names of the list and settled on the rather long ones with underscores, but I can also understand the argument for keeping them short but replacing the output with the longer full names in the print method. So `cond`, `zi`, and `disp` are good middle grounds IMO. 

The `pscl` package uses `count` and `zero` consistently for naming the two parts of the model. I think `cond` and `zi` are more accurate descriptions with similar length, so let's go with that.

`pscl` has the following print formats comparable to our `fixef`
```
> coef(fm_zip)
count_(Intercept)    count_femWomen  count_marMarried        count_kid5         count_phd        count_ment  zero_(Intercept) 
      0.553994991      -0.231609013       0.131971483      -0.170473724       0.002525874       0.021542730      -1.681350337 
> coef(fm_zip, "count")
 (Intercept)     femWomen   marMarried         kid5          phd         ment 
 0.553994991 -0.231609013  0.131971483 -0.170473724  0.002525874  0.021542730 
> coef(fm_zip, "zero")
(Intercept) 
   -1.68135 
```

**Comment by mebrooks** [comment]
On the topic of DRY...Koen and I were brainstorming ways to obey DRY and that's when I said "I think this is taking longer than it should". It felt like avoiding repeating 4 lines of code wasn't worth the time it was taking.

**Comment by bbolker** [comment]
the `pscl` format is interesting (and it's certainly worth considering alternatives).  I think it's easier to convert from the list to the flattened format than vice versa:

```
bb <- list(cond=c(a=1,b=2),zi=c(a=1,b=1))
list_to_vec <- function(bb) {
   tmpf <- function(x,y) setNames(y,paste(x,names(y),sep="_"))
   unlist(unname(Map(tmpf,names(bb),bb)))
}
vec_to_list <- function(cc) {
   prefix <- gsub("_[^_]+$","",names(cc))
   res <- lapply(split(cc,prefix),
          function(x) setNames(x,gsub("^[^_]+_","",names(x))))
   setNames(res,unique(prefix))
}
## round trip
identical(bb,vec_to_list(list_to_vec(bb)))
```

We could try to figure out a way to offer these helper functions to
the user ... one quick-and-dirty way is in the help page examples

**Comment by koenvanbenthem** [comment]
The DRY way:

print.fixef.glmmTMB<-function(object, digits = max(3, getOption("digits") -
3)){
  name <- list(cond = "Conditional model", zi = "Zero inflation", disp =
"Dispersion")
  for(x in names(object)){
    if(length(object[[x]]>0)){
      cat("\n",name[[x]],"\n",sep="")
      print.default(format(object[[x]],digits=digits), print.gap = 2L,
quote = FALSE)
    }
  }
}

However, the problem is that in this case it would also print the
dispersoin if it has just one parameter (intercept only) -- last week we
discussed that this was not desirable. To deal with that I could
incorporate an if-statement (but then we are already quite close to just
splitting it up and writing separate functions for each case..)

2015-08-25 19:54 GMT+02:00 Ben Bolker <notifications@github.com>:

**Comment by koenvanbenthem** [comment]
DRY (corrected parenthesis and included if-statement for dispersion
intercept):

print.fixef.glmmTMB<-function(object, digits = max(3, getOption("digits") -
3)){
  name <- list(cond = "Conditional model", zi = "Zero inflation", disp =
"Dispersion")
  for(x in names(object)){
    if((length(object[[x]])-as.numeric(x == 'disp' & '(Intercept)' %in%
names(object[[x]])))>0){
      cat("\n",name[[x]],"\n",sep="")
      print.default(format(object[[x]],digits=digits), print.gap = 2L,
quote = FALSE)
    }
  }
}

2015-08-25 20:29 GMT+02:00 Koen van Benthem <koenvanbenthem@gmail.com>:

**Comment by arnima-github** [comment]
Hi guys, thanks for a great development sprint in Zürich last week. I just 
returned from 3 days of trout fishing in the Icelandic highlands.

I agree with the points raised by Martin and the discussion that followed, 
regarding shorter element names and not hiding the inner structure 
completely from the user.

After finishing the ranef() method over the weekend, I pushed it to 
Github. The crux of it was to adapt ranef.merMod() into an inner arrange() 
function that our ranef.glmmTMB() calls twice. I also made minor 
modifications to fixef() just to align it a bit with the ranef() design, 
which I documented on the ranef() roxygenized help page. I can see things 
have been improved since then and I'm happy with all the changes.

Arni
